File: packages/rust/src/remote_process_client.rs
use serde::{Deserialize, Serialize};
use std::io::{self, BufReader, BufWriter};
use std::net::TcpStream;

pub struct RemoteProcessClient {
    buffer: String,
    reader: BufReader<TcpStream>,
    writer: BufWriter<TcpStream>,
}

impl RemoteProcessClient {
    pub fn connect(host: &str, port: u16) -> io::Result<Self> {
        let stream = TcpStream::connect((host, port))?;
        stream.set_nodelay(true)?;
        let stream_clone = stream.try_clone()?;
        let mut result = RemoteProcessClient {
            buffer: String::new(),
            reader: BufReader::new(stream),
            writer: BufWriter::new(stream_clone),
        };
        use std::io::Write;
        writeln!(result.writer, "json")?;
        result.writer.flush()?;
        Ok(result)
    }

    pub fn read<T: for<'a> Deserialize<'a>>(&mut self) -> io::Result<Option<T>> {
        use std::io::BufRead;
        self.buffer.clear();
        self.reader.read_line(&mut self.buffer)?;
        trace!("Read line: {:?}", self.buffer);
        if self.buffer.is_empty() {
            return Ok(None);
        }
        Ok(Some(::serde_json::from_str(&self.buffer)?))
    }
    pub fn write<T: Serialize>(&mut self, value: &T) -> io::Result<()> {
        use std::io::Write;
        writeln!(self.writer, "{}", ::serde_json::to_string(value)?)?;
        self.writer.flush()?;
        Ok(())
    }
    pub fn write_token(&mut self, token: &str) -> io::Result<()> {
        use std::io::Write;
        writeln!(self.writer, "{}", token)?;
        self.writer.flush()?;
        Ok(())
    }
}

File: packages/javascript/remote-process-client.js
/**
 * Created by Quake on 18.12.2018
 */
'use strict';

const Game = require('./model/Game');
const Action = require('./model/Action');
const Rules = require('./model/Rules');

let client = null;
let strBuffer = [];

module.exports.connect = function connect(host, port, onConnect) {
    const net = require('net');
    var request = [];
    var busy;

    client = net.connect(port, host, function connectHandler() {
        // 'connect' listener
        console.log('connected to server!');
        writeline("json");
        onConnect();
    });
    client.setNoDelay();

    client.on('data', dataHandler);
    client.on('error', function onError(e) {
        console.log("SOCKET ERROR: " + e.message);
        process.exit(1);
    });

    client.on('close', function onClose() {
        console.log('server closed connection.');
        client.unref();
        process.exit();
    });
    client.on('end', function onEnd() {
        console.log('disconnected from server');
        process.exit();
    });

    let tickData = '';
    function dataHandler(data) {
        busy = true;
        if (data) {
            //by Gondragos
            // sometimes we get tick data by parts 
            tickData = tickData + data.toString();
            // tickData is complete when we get '\n'
            if (tickData[tickData.length -1] === '\n') {
                strBuffer.push(tickData);
                tickData = '';
            }
        }
        while (request.length && strBuffer.length) {
            let cb = request.shift();
            let data = strBuffer.shift();
            cb(data);
        }
        busy = false;
    }

    function readline() {
        return strBuffer.shift();
    }
    
    function writeline(line) {
        line += "\n";
        client.write(line);
    }

    function readSequence(cb) {
        if (!cb) {
            throw 'Callback expected';
        }
        request.push(cb);
        if (!busy) {
            dataHandler();
        }
    }
    
    this.readRules = function(cb) {
        readSequence(function(data) {
            let rules = new Rules();
            rules.read(data);
            
            cb(rules);
        });
    }
    
    this.readGame = function(cb) {
        readSequence(function(data) {
            let game = new Game();
            game.read(data);
            
            cb(game);
        });
    }
    
    this.write = function(actions, customRendering) {
        let map = {};
        for (let id in actions) {
            map[id] = actions[id].toObject();
        }
        let json = JSON.stringify(map);
        writeline(json + '|' + customRendering + '\n<end>');
    }
    
    this.writeTokenMessage = function(token, cb) {
        writeline(token);
        cb();
    }
}